controller/dataplane/custom_plugin.go

```diff
@@ -0,0 +1,87 @@
+package dataplane
+
+import (
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/gateway-operator/pkg/consts"
+	k8sresources "github.com/kong/gateway-operator/pkg/utils/kubernetes/resources"
+)
+
+type customPlugin struct {
+	// Name of the KongPluginInstallation resource.
+	Name string
+	// ConfigMapNN is the namespace/name of the ConfigMap that contains the plugin.
+	ConfigMapNN types.NamespacedName
+	// Generation is the generation of the KongPluginInstallation that contains the plugin.
+	Generation int64
+}
+
+func withCustomPlugins(customPlugins []customPlugin) k8sresources.DeploymentOpt {
+	// Noop/cleanup operation that is safe to execute if no plugins are provided.
+	if len(customPlugins) == 0 {
+		return func(d *appsv1.Deployment) {
+			// It's safe to perform a delete operation on a nil map.
+			delete(
+				d.Spec.Template.Annotations,
+				consts.AnnotationKongPluginInstallationGenerationInternal,
+			)
+		}
+	}
+
+	var (
+		kpisNames        = make([]string, 0, len(customPlugins)+1)
+		kpisGenerations  = make([]string, 0, len(customPlugins))
+		kpisVolumeMounts = make([]corev1.VolumeMount, 0, len(customPlugins))
+		kpisVolumes      = make([]corev1.Volume, 0, len(customPlugins))
+	)
+	kpisNames = append(kpisNames, "bundled") // To preserve plugins available in the default image.
```

Comment:
How about extracting `bundled` as a constant and documenting this, explaining it is a special value Kong accepts in its `KONG_PLUGINS` to use the default builtin plugins bundle?

Reply:
I'm proposing an alternative approach, I hope you'll like it